src/charm.py

```diff
@@ -213,6 +278,142 @@ def _on_get_rule_action(self, event: ActionEvent) -> None:
+        for rule_type in ("allow", "deny"):
+            rules = self._render_access_rules(
+                rule_type=rule_type,
+                protected_urls=event.protected_urls,
+                allowed_endpoints=event.allowed_endpoints,
+                relation_app_name=event.relation_app_name,
+            )
+
+            if rules:
+                # Push the rules to the container
+                filename = f"{self._oathkeeper_access_rules_path}/access-rules-{event.relation_app_name}-{rule_type}.json"
+                self._container.push(filename, str(rules), make_dirs=True)
+                access_rules_locations.append(filename)
+
+        self._set_auth_proxy_relation_peer_data(
+            event.relation_id, dict(access_rules_locations=access_rules_locations)
+        )
+
+        self._push_oathkeeper_config()
+
+    def _rule_template(
+        self, rule_id: str, url: str, authenticator: str, mutator: str, error_handler: str
+    ) -> Dict:
+        return {
+            "id": rule_id,
+            "match": {"url": url, "methods": ["GET", "POST"]},
+            "authenticators": [{"handler": authenticator}],
+            "mutators": [{"handler": mutator}],
+            "authorizer": {"handler": "allow"},
+            "errors": [{"handler": error_handler}],
+        }
+
+    def _render_access_rules(
+        self,
+        rule_type: str,
+        protected_urls: List[str],
+        allowed_endpoints: List[str],
+        relation_app_name: str,
+    ) -> Optional[List[Dict]]:
+        """Render access rules from a template."""
+        rules = list()
+
+        for url_index, url in enumerate(protected_urls):
+            if url.startswith("https"):
+                # Replace with regex to match both http and https
+                url = url.replace("https", "<https|http>")
+
+            if rule_type == "allow":
+                if not allowed_endpoints:
+                    return None
+                for endpoint in allowed_endpoints:
+                    allow_rule = self._rule_template(
+                        rule_id=f"{relation_app_name}:{endpoint}:{url_index}:allow",
+                        url=f"{url}/{endpoint}",
+                        authenticator="noop",
+                        mutator="noop",
+                        error_handler="json",
+                    )
+
+                    rules.append(allow_rule)
+
+            if rule_type == "deny":
+                if allowed_endpoints:
+                    # Render a regex to exclude allowed endpoints
+                    exclude_endpoints = list()
+                    for endpoint in allowed_endpoints:
+                        if "/" in endpoint:
+                            # about/app must become a group: (about/app)
+                            endpoint = f"({endpoint})"
+                        exclude_endpoints.append(endpoint)
+
+                    # Add | alternation
+                    deny_regex = f"{url}/<(?!{'|'.join(exclude_endpoints)}).*>"
```

Comment:
With this regex we catch all endpoints that start with path. 
If someone wanted to allow access to `http://localhost/home`, we would create this rule:
`http://localhost/<(?!home).*>`
BUT if there is a path`http://localhost/home-protected`, which they wanted to protect, it would not be protected.

I think that we should add to all excluded endpoints the prefix `^` and the suffix `$`. So the regex would be:
`^http://localhost/<(?!home$).*>`

Reply:
Thanks for catching that, I changed the regex